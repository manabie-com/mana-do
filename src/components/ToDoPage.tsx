import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import RLDD from "react-list-drag-and-drop/lib/RLDD";
import styled from "styled-components";
import {
  createTodo,
  toggleAllTodos,
  deleteAllTodos,
  updateTodoStatus,
  deleteTodo,
  setTodos,
  updateTodo,
} from "../store/actions";
import Service from "../service";
import { TodoStatus, EnhanceTodoStatus, TodoList } from "../models/todo";
import { AppState } from "../store/reducer";
import { LIST_TAB } from "../utils/constants";
import { isValidCreate } from "../utils/validate";
import Modal from "../components/Modal";

const Container = styled.div`
  border: 1px solid rgba(0, 0, 0, 0.13);
  border-radius: 8px;
  width: 500px;
  margin-top: 5rem;
  padding: 24px;
  box-shadow: 2px 2px 1px rgba(0, 0, 0, 0.09), 3px 2px 3px rgba(0, 0, 0, 0.05);
`;

const ToDoPage = () => {
  const todos = useSelector((state: AppState) => state.todos);
  const dispatch = useDispatch();
  const [showing, setShowing] = useState<EnhanceTodoStatus>("ALL");
  const [items, setItems] = useState<TodoList[]>([]);
  const [input, setInput] = useState<string>("");
  const [itemEdit, setItemEdit] = useState<TodoList>();
  const [inputItem, setInputItem] = useState<string>("");
  const [visible, setVisible] = useState<boolean>(false);

  const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter" && isValidCreate(todos, input)) {
      const resp = await Service.createTodo(input);
      if (resp.id) {
        dispatch(createTodo(resp));
        setInput("");
      }
    }
  };

  const onUpdateTodo = async (
    e: React.KeyboardEvent<HTMLInputElement>,
    id: string
  ) => {
    if (e.key === "Enter" && isValidCreate(todos, inputItem)) {
      dispatch(updateTodo(id, inputItem));
      setItemEdit(undefined);
      setInputItem("");
      setVisible(false);
    }
  };

  useEffect(() => {
    var listData = [] as TodoList[];
    todos
      .filter((item) => showing === "ALL" || item.status === showing)
      .forEach((item) => {
        listData.push({
          ...item,
          id: item.number_id,
          todo_id: item.id,
        });
      });
    setItems(listData);
  }, [todos, showing]);

  const onUpdateTodoStatus = (
    e: React.ChangeEvent<HTMLInputElement>,
    todoId: string
  ) => {
    dispatch(updateTodoStatus(todoId, e.target.checked));
  };

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(toggleAllTodos(e.target.checked));
  };

  const onDeleteAllTodo = () => {
    dispatch(deleteAllTodos());
  };

  const onDeleteTodo = (todoId: string) => {
    dispatch(deleteTodo(todoId));
  };

  const itemRenderer = (item: TodoList): any => {
    return (
      <div key={item.id} className="ToDo__item">
        <input
          type="checkbox"
          checked={item.status === TodoStatus.COMPLETED}
          onChange={(e) => onUpdateTodoStatus(e, item.todo_id)}
        />
        <span
          onClick={() => {
            setItemEdit(item);
            setVisible(true);
            setInputItem(item.content);
          }}
        >
          {item.content}
        </span>
        <input
          type="button"
          className="Todo__delete"
          onClick={() => onDeleteTodo(item.todo_id)}
          value="X"
        />
      </div>
    );
  };
  const handleRLDDChange = (reorderedItems: Array<TodoList>) => {
    dispatch(
      setTodos(
        reorderedItems.map((item) => {
          const newItem = {
            ...item,
            id: item.todo_id,
            number_id: item.id,
          };
          return newItem;
        })
      )
    );
  };

  return (
    <Container>
      <div className="Todo__creation">
        <input
          value={input}
          className="Todo__input"
          placeholder="What need to be done?"
          onKeyDown={onCreateTodo}
          onChange={(e) => setInput(e.currentTarget.value)}
        />
      </div>
      <div className="Todo__toolbar">
        <input
          style={{ visibility: items.length > 0 ? "visible" : "hidden" }}
          type="checkbox"
          checked={
            items.length ===
            items.filter((item) => item.status === TodoStatus.COMPLETED).length
          }
          onChange={onToggleAllTodo}
        />
        <div className="Todo__tabs">
          {LIST_TAB.map((tab) => {
            return (
              <button
                key={tab.value}
                className={`Action__btn ${showing === tab.value && "active"}`}
                onClick={() => setShowing(tab.value)}
              >
                {tab.text}
              </button>
            );
          })}
        </div>
        <button className="Action__btn clear__all" onClick={onDeleteAllTodo}>
          Clear all todos
        </button>
      </div>
      <div className="ToDo__list">
        {showing === "ALL" ? (
          <RLDD
            items={items}
            itemRenderer={itemRenderer}
            onChange={handleRLDDChange}
          />
        ) : (
          items.map((item) => itemRenderer(item))
        )}
      </div>
      <Modal
        visible={visible}
        title="Edit todo"
        onClose={(e) => setVisible(false)}
      >
        <input
          className="Todo__input"
          value={inputItem}
          onKeyDown={(e) => onUpdateTodo(e, itemEdit?.todo_id || "")}
          onChange={(e) => {
            setInputItem(e.currentTarget.value);
          }}
        />
      </Modal>
    </Container>
  );
};

export default ToDoPage;
