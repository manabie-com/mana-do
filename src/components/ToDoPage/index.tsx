export { default } from './ToDoPage';
export * from './ToDoPage';
