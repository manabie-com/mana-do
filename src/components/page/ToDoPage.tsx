import React, { useEffect, useState } from 'react';

import { setTodos, toggleAllTodos, deleteAllTodos } from '../../store/actions';
import Service from '../../service';
import { TodoStatus } from '../../models/todo';
import { isTodoCompleted } from '../../utils';
import ToDoInput from '../common/ToDoInput/ToDoInput.components';
import ToDoItem from '../common/ToDoItem/ToDoItem.components';
import ActionButton from '../common/ActionButton/ActionButton.components';
import ToDoTabs from '../common/ToDoToolbar/ToDoTabs.components';
import { useToDoPageContext } from '../../context/ToDoPageProvider';

export type EnhanceTodoStatus = TodoStatus | 'ALL';

const ToDoPage = () => {
  const { todos, dispatch } = useToDoPageContext();
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  // create error message state for error handling
  const [error, setError] = React.useState<String>('');
  useEffect(() => {
    (async () => {
      try {
        const resp = await Service.getTodos();
        dispatch(setTodos(resp));
      } catch (error) {
        setError('Have an error when getTodos');
      }
    })();
  }, [dispatch]);

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { checked } = e.target;
    Service.toggleAllTodos(checked);
    dispatch(toggleAllTodos(checked));
  };

  const onDeleteAllTodo = async () => {
    await Service.deleteAllTodos();
    dispatch(deleteAllTodos());
  };

  const showTodos = todos.filter((todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const activeTodos = todos.reduce(function (accum, todo) {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);

  return (
    <div className='ToDo__container'>
      <div className='Todo__creation'>
        <ToDoInput />
      </div>
      <div className='ToDo__list'>
        {showTodos.map((todo) => {
          const { id, content } = todo;
          return (
            <ToDoItem
              key={id}
              id={id}
              checked={isTodoCompleted(todo)}
              content={content}
            />
          );
        })}
      </div>
      <div className='Todo__toolbar'>
        {todos.length > 0 ? (
          <input
            data-testid='toggle-all-todos'
            type='checkbox'
            checked={activeTodos === 0}
            onChange={onToggleAllTodo}
          />
        ) : (
          <div />
        )}
        <ToDoTabs setShowing={setShowing} />
        <ActionButton
          data-testid='clear-all-todos'
          className='Action__btn'
          onClick={onDeleteAllTodo}
        >
          Clear all todos
        </ActionButton>
      </div>
      {error && (
        <div>
          <p style={{ color: 'red' }}>{error}</p>
        </div>
      )}
    </div>
  );
};

export default ToDoPage;
