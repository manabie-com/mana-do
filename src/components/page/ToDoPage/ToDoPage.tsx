import React, { useEffect, useState } from 'react';

import {
  setTodos,
  toggleAllTodos,
  deleteAllTodos,
} from '../../../store/actions';
import Service from '../../../service';
import { TodoStatus } from '../../../models/todo';
import { isTodoCompleted } from '../../../utils';
import ToDoInput from '../../common/ToDoInput/ToDoInput.components';
import ToDoItem from '../../common/ToDoItem/ToDoItem.components';
import ActionButton from '../../common/ActionButton/ActionButton.components';
import ToDoTabs from '../../common/ToDoTabs/ToDoTabs.components';
import { useToDoPageContext } from '../../../context/ToDoPageProvider';

export type EnhanceTodoStatus = TodoStatus | 'ALL';

const ToDoPage = () => {
  const { todos, dispatch } = useToDoPageContext();
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  // create error message state for error handling
  const [error, setError] = React.useState<string>('');
  const [loading, setLoading] = React.useState<boolean>(false);
  useEffect(() => {
    setLoading(true);
    (async () => {
      try {
        const resp = await Service.getTodos();
        dispatch(setTodos(resp));
      } catch (error) {
        setError('Have an error when getTodos');
      }
    })();
    setLoading(false);
  }, [dispatch]);

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { checked } = e.target;
    // Optimis update
    Service.toggleAllTodos(checked);
    dispatch(toggleAllTodos(checked));
  };

  const onDeleteAllTodo = async () => {
    try {
      await Service.deleteAllTodos();
      dispatch(deleteAllTodos());
    } catch (error) {
      setError('Have error when delete all todos');
    }
  };

  const showTodos = todos.filter((todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const activeTodos = todos.reduce(function (accum, todo) {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);

  return (
    <div className='ToDo__container'>
      <div className='Todo__creation'>
        <ToDoInput />
      </div>
      {!loading ? (
        <>
          <div className='ToDo__list'>
            {!showTodos.length && <span>Nothing to do</span>}
            {showTodos.map((todo) => {
              const { id, content, created_date } = todo;
              return (
                <ToDoItem
                  key={id}
                  id={id}
                  checked={isTodoCompleted(todo)}
                  content={content}
                  createdDate={new Date(created_date)}
                />
              );
            })}
          </div>
          <div className='Todo__toolbar'>
            {todos.length > 0 ? (
              <input
                data-testid='toggle-all-todos'
                type='checkbox'
                checked={activeTodos === 0}
                onChange={onToggleAllTodo}
              />
            ) : (
              <div />
            )}
            <ToDoTabs setShowing={setShowing} activeTab={showing} />
            <ActionButton
              data-testid='clear-all-todos'
              className='Action__btn'
              onClick={onDeleteAllTodo}
            >
              Clear all todos
            </ActionButton>
          </div>
          {error && (
            <div>
              <p style={{ color: 'red' }}>{error}</p>
            </div>
          )}
        </>
      ) : (
        <span>Loading</span>
      )}
    </div>
  );
};

export default ToDoPage;
