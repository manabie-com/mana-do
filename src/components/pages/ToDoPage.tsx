import React, { useEffect, useReducer, useRef, useState } from 'react';

import reducer, { initialState } from '../../store/reducer';
import {
    setTodos,
    createTodo,
    toggleAllTodos,
    deleteAllTodos,
    updateTodoStatus,
    deleteTodo
} from '../../store/actions';
import Service from '../../service';
import { TodoStatus } from '../../models/todo';
import ToolBar from '../common/tool-bar/ToolBar';
import TaskCreateInput from '../common/to-do-create-input/ToDoCreateInput';
import ToDoList from '../common/to-do-list/ToDoList';

type EnhanceTodoStatus = TodoStatus | 'ALL';


const ToDoPage = () => {
    const [{ todos }, dispatch] = useReducer(reducer, initialState);
    const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
    const inputRef = useRef<any>(null);

    useEffect(() => {
        (async () => {
            const resp = await Service.getTodos();

            dispatch(setTodos(resp || []));
        })()
    }, [])

    const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === 'Enter') {
            const resp = await Service.createTodo(inputRef.current.value);
            dispatch(createTodo(resp));
        }
    }

    const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: any) => {
        dispatch(updateTodoStatus(todoId, e.target.checked))
    }

    const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(toggleAllTodos(e.target.checked))
    }

    const onDeleteTodo = (todoId: string) => {
        dispatch(deleteTodo(todoId));
    }

    const onDeleteAllTodo = () => {
        dispatch(deleteAllTodos());
    }
    const toolBarProps = {
        todos,
        onToggleAllTodo,
        setShowing,
        onDeleteAllTodo
    }
    const taskCreateInputProps = {
        inputRef,
        onCreateTodo,
    }
    const toDoListProps = {
        todos,
        onUpdateTodoStatus,
        onDeleteTodo,
        showing
    }
    return (
        <div className="to-do__container">
            <div className="to-do__creation">
                <TaskCreateInput {...taskCreateInputProps} />
            </div>
            <div className="to-do__toolbar">
                <ToolBar {...toolBarProps} />
            </div>
            <div className="to-do__list">
                <ToDoList {...toDoListProps} />
            </div>
        </div>
    );
};

export default ToDoPage;