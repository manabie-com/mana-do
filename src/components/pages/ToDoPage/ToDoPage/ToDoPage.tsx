import React, { useEffect, useReducer, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { TodoStatus } from '../../../../models/todo';
import Service from '../../../../service';
import { setTodos } from '../../../../store/actions';
import reducer, { initialState } from '../../../../store/reducer';
import CreateTask from '../CreateTask/CreateTask';
import ShowTodoList from '../ShowTodoList/ShowTodoList';
import TodoToolBar from '../TodoToolBar/TodoToolBar';

import './toDoPage.scss'

export type EnhanceTodoStatus = TodoStatus | 'ALL';


const ToDoPage = () => {
    const [{ todos }, dispatch] = useReducer(reducer, initialState);
    const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');

    const history = useHistory()
    useEffect(() => {
        if (!localStorage.getItem('token')) {
            alert('Vui lòng đăng nhập để tiếp tục!')
            history.push('')
        }
        const data = localStorage.getItem('task')

        if (data) {
            dispatch(setTodos(JSON.parse(data)))
        }

        else {
            (async () => {
                const resp = await Service.getTodos();
                dispatch(setTodos(resp || []));
            })()
        }
    }, [])

    return (
        <div className="ToDo__container">
            <CreateTask dispatch={dispatch} todos={todos} />
            <ShowTodoList showing={showing} todos={todos} dispatch={dispatch} />
            <TodoToolBar setShowing={setShowing} todos={todos} dispatch={dispatch} />
        </div>
    );
};

export default ToDoPage;