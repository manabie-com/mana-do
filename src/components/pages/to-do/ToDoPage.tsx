import { useEffect, useReducer, useState } from "react";

import reducer, { initialState } from "../../../store/reducer";
import { setTodos } from "../../../store/actions";
import Service from "../../../service";
import { TodoStatus } from "../../../models/todo";
import ToolBar from "../../common/tool-bar/ToolBar";
import TaskCreateInput from "../../common/to-do-create-input/ToDoCreateInput";
import ToDoList from "../../common/to-do-list/ToDoList";

type EnhanceTodoStatus = TodoStatus | "ALL";

const ToDoPage = () => {
    const token = localStorage.getItem('token');
    if (!token) {
        window.location.href = 'login';
    }
    const [{ todos }, dispatch] = useReducer(reducer, initialState);
    const [showing, setShowing] = useState<EnhanceTodoStatus>("ALL");

    useEffect(() => {
        (async () => {
            const res = await Service.getTodos();
            dispatch(setTodos(res.data || []));
        })();
    }, []);

    const toolBarProps = {
        todos,
        dispatch,
        setShowing,
    };
    const taskCreateInputProps = {
        dispatch,
    };
    const toDoListProps = {
        todos,
        dispatch,
        showing,
    };
    return (
        <div className="to-do__container">
            <div className="to-do__creation">
                <TaskCreateInput {...taskCreateInputProps} />
            </div>
            <div className="to-do__toolbar">
                <ToolBar {...toolBarProps} />
            </div>
            <div className="to-do__list">
                <ToDoList {...toDoListProps} />
            </div>
        </div>
    );
};

export default ToDoPage;
