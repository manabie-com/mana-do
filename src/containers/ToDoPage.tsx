import { Checkbox } from "components/checkbox";
import { Guide } from "components/guide";
import { Header } from "components/header";
import { Input } from "components/input";
import { TodoItemList } from "components/todoItemList";
import { TodoTab } from "components/todoTab";
import { useStateReducer } from "hooks";
import { TodoStatus } from "models";
import React, { useCallback, useMemo, useRef, useState } from "react";
import { toast, ToastContainer } from "react-toastify";
import Service from "service";
import {
  createTodo,
  deleteAllTodos,
  toggleAllTodos,
} from "store/action-handlers";
import { isTodoActive, isTodoCompleted } from "utils";
import "./styles.scss";

export type EnhanceTodoStatus = TodoStatus | "ALL";

const ToDoPage = () => {
  const [{ todos }, dispatch] = useStateReducer();
  const [showing, setShowing] = useState<EnhanceTodoStatus>("ALL");
  const inputRef = useRef<any>(null);

  /** Event hanlders --- start */

  const onCreateTodo = useCallback(
    async (e: React.KeyboardEvent<HTMLInputElement>): Promise<void> => {
      if (e.key === "Enter" && inputRef.current) {
        const { value } = inputRef.current;
        if (!value) {
          return;
        }
        try {
          const resp = await Service.createTodo(value);
          dispatch(createTodo(resp));
          inputRef.current.value = "";
        } catch (error) {
          toast.error("Something went wrong!");
        }
      }
    },
    [dispatch]
  );

  const onToggleAllTodo = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>): void => {
      dispatch(toggleAllTodos(e.target.checked));
    },
    [dispatch]
  );

  const onDeleteAllTodo = useCallback(() => {
    dispatch(deleteAllTodos());
  }, [dispatch]);

  /** Event hanlders --- end */

  const todoTasks = useMemo(
    () =>
      todos.filter((todo) => {
        switch (showing) {
          case TodoStatus.ACTIVE:
            return isTodoActive(todo.status);
          case TodoStatus.COMPLETED:
            return isTodoCompleted(todo.status);
          default:
            return true;
        }
      }),
    [showing, todos]
  );

  return (
    <div className="todo">
      <div className="todo__container">
        <Header />
        <div className="todo__creation">
          <Input
            ref={inputRef}
            placeholder="What need to be done?"
            name="todo"
            onKeyDown={onCreateTodo}
          />
        </div>
        <TodoItemList todos={todoTasks} dispatch={dispatch} />
        <div className="todo__toolbar">
          {todoTasks.length > 0 ? (
            <Checkbox onChange={onToggleAllTodo} data-testid="checkbox-all" />
          ) : (
            <div />
          )}
          <TodoTab
            selectedStatus={showing}
            onClickSelectStatus={setShowing}
            onDeleteAllTodo={onDeleteAllTodo}
          />
        </div>
      </div>
      <ToastContainer />
      <Guide />
    </div>
  );
};

export default ToDoPage;
