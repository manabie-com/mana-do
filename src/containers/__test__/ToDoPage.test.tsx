import { fireEvent, render } from "@testing-library/react";
import ToDoPage from "containers/ToDoPage";
import React from "react";
import { getElementByTestId } from "utils/testUtils";

const renderComponent = () => {
  return render(<ToDoPage />);
};

describe("<ToDoPage /> rendering", () => {
  it("should match snapshot", () => {
    const wrapper = renderComponent();
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should render UI correctly", () => {
    renderComponent();

    const inputElement = getElementByTestId("input") as HTMLInputElement;

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute(
      "placeholder",
      "What need to be done?"
    );
  });
});

describe("<ToDoPage /> interacting", () => {
  it("should be able to type input", () => {
    const defaultValue = "Doing excercises";

    renderComponent();
    const inputElement = getElementByTestId("input") as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: defaultValue } });
    expect(inputElement.value).toBe(defaultValue);
  });
});
