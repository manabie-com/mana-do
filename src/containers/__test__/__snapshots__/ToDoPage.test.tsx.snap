// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToDoPage /> rendering should match snapshot 1`] = `
<div>
  <div
    class="todo"
  >
    <div
      class="todo__container"
    >
      <div
        class="header"
      >
        <svg
          color="#3A52D1"
          fill="currentColor"
          font-size="36"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(58, 82, 209);"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M22 5.18L10.59 16.6l-4.24-4.24 1.41-1.41 2.83 2.83 10-10L22 5.18zm-2.21 5.04c.13.57.21 1.17.21 1.78 0 4.42-3.58 8-8 8s-8-3.58-8-8 3.58-8 8-8c1.58 0 3.04.46 4.28 1.25l1.44-1.44A9.9 9.9 0 0012 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10c0-1.19-.22-2.33-.6-3.39l-1.61 1.61z"
          />
        </svg>
        <h1
          class="header__title"
          data-testid="header"
        >
          MANA DO
        </h1>
      </div>
      <div
        class="todo__creation"
      >
        <input
          class="input-todo "
          data-testid="input"
          id=""
          name="todo"
          placeholder="What need to be done?"
          type="text"
        />
      </div>
      <div
        class="list-todo"
      />
      <div
        class="todo__toolbar"
      >
        <div />
        <div
          class="todo-tab"
          data-testid="todo-tab"
        >
          <button
            aria-label="ALL"
            class="btn-todo btn-todo--main btn-todo--selected"
            data-testid="button"
          >
            ALL
          </button>
          <button
            aria-label="ACTIVE"
            class="btn-todo btn-todo--main "
            data-testid="button"
          >
            ACTIVE
          </button>
          <button
            aria-label="COMPLETED"
            class="btn-todo btn-todo--main "
            data-testid="button"
          >
            COMPLETED
          </button>
        </div>
        <button
          aria-label="Clear all"
          class="btn-todo btn-todo--danger"
          data-testid="button"
        >
          Clear all todos
        </button>
      </div>
    </div>
    <div
      class="Toastify"
    />
    <div
      class="guide"
    >
      <i
        class="guide__text"
      >
        * Double-click to the label to edit the corresponding task!
      </i>
    </div>
  </div>
</div>
`;
